src/common/utils/pagination.ts

```diff
@@ -0,0 +1,55 @@
+import { List, PaginationOptions } from '../interfaces';
+
+export class AutoPaginatable<T> {
+  readonly object: 'list' = 'list';
+  readonly options: PaginationOptions;
+
+  constructor(
+    private list: List<T>,
+    private apiCall: (params: PaginationOptions) => Promise<List<T>>,
+    options?: PaginationOptions,
+  ) {
+    this.options = {
+      ...options,
+    };
+  }
+
+  get data(): T[] {
+    return this.list.data;
+  }
+
+  get listMetadata() {
+    return this.list.listMetadata;
+  }
+
+  private async *generatePages(params: PaginationOptions): AsyncGenerator<T[]> {
+    const result = await this.apiCall({
+      ...this.options,
+      after: params.after,
+    });
+
+    yield result.data;
+
+    if (result.listMetadata.after) {
+      yield* this.generatePages({ after: result.listMetadata.after });
```

Comment:
Related to [Marshall's previous comment](https://github.com/workos/workos-node/pull/788#discussion_r1288869625): Is this also a recursive call that would impact stack safety? If so, we could refactor this to use to use a `while` or `do while`.

I'm not sure how much it matters in practice though, so just noting it.